File: python/atelier_bench/services/profit_margin_service.py
"""profit-margin service."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..support.abstract_service import AbstractService

if TYPE_CHECKING:
    from ..dataset import Dataset


class ProfitMarginService(AbstractService):
    NAME = "profit-margin"

    def __init__(self) -> None:
        super().__init__(self.NAME)

    def ratio(self, data: "Dataset") -> float:
        revenue = data.revenue_cents()
        return 0.0 if revenue == 0 else (revenue - data.parts_cost_cents()) / revenue

File: go/atelier/services/profit_margin_service.go
// Package services carries the 12 services. Note the name: core/services is a
// DIFFERENT package with the same name, and a file needing both aliases one.
package services

import (
	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/dataset"
	"atelier.example/lane/atelier/support"
)

// Compile-time proof of satisfaction.
var _ contracts.ServiceContract = (*ProfitMarginService)(nil)

// ProfitMarginService answers the profit-margin questions.
type ProfitMarginService struct {
	support.BaseService
}

// NewProfitMarginService builds the service.
func NewProfitMarginService() *ProfitMarginService {
	return &ProfitMarginService{BaseService: support.NewBaseService("profit-margin")}
}

// Ratio is gross profit over revenue.
func (s *ProfitMarginService) Ratio(data *dataset.Dataset) float64 {
	revenue := data.RevenueCents()
	if revenue == 0 {
		return 0
	}
	return float64(revenue-data.PartsCostCents()) / float64(revenue)
}

// Run renders the service result as report rows.
func (s *ProfitMarginService) Run(data *dataset.Dataset) []contracts.ReportRow {
	return []contracts.ReportRow{contracts.Row("ratio", s.Ratio(data))}
}
